finality-aleph/src/validator_network/protocol_negotiation.rs

```diff
@@ -0,0 +1,267 @@
+        }
+    }
+
+    #[tokio::test]
+    async fn fails_when_no_intersection() {
+        let (stream1, stream2) = duplex(4096);
+        let mut too_high_protocol_range = supported_protocol_range();
+        too_high_protocol_range.0 = too_high_protocol_range.1 + 1;
+        too_high_protocol_range.1 = too_high_protocol_range.0 + 1;
+        let negotiation1 = negotiate_protocol_version(stream1, supported_protocol_range()).fuse();
+        pin_mut!(negotiation1);
+        let negotiation2 =
+            negotiate_protocol_version(stream2, too_high_protocol_range.clone()).fuse();
+        pin_mut!(negotiation2);
+        for _ in 0..2 {
+            tokio::select! {
+                result = &mut negotiation1 => incorrect_negotiation(result, ProtocolNegotiationError::ProtocolMismatch(supported_protocol_range(), too_high_protocol_range.clone())),
+                result = &mut negotiation2 => incorrect_negotiation(result, ProtocolNegotiationError::ProtocolMismatch(too_high_protocol_range.clone(), supported_protocol_range())),
+            }
+        }
+    }
+
+    #[tokio::test]
+    async fn fails_when_bad_negotiation() {
+        let (stream1, stream2) = duplex(4096);
+        let mut too_high_protocol_range = supported_protocol_range();
+        too_high_protocol_range.0 = too_high_protocol_range.1 + 1;
+        too_high_protocol_range.1 = too_high_protocol_range.0 + 1;
+        let negotiation1 =
+            negotiate_protocol_version(stream1, too_high_protocol_range.clone()).fuse();
+        pin_mut!(negotiation1);
+        let negotiation2 =
+            negotiate_protocol_version(stream2, too_high_protocol_range.clone()).fuse();
+        pin_mut!(negotiation2);
+        for _ in 0..2 {
+            tokio::select! {
+                result = &mut negotiation1 => incorrect_negotiation(result, ProtocolNegotiationError::BadChoice(too_high_protocol_range.1)),
+                result = &mut negotiation2 => incorrect_negotiation(result, ProtocolNegotiationError::BadChoice(too_high_protocol_range.1)),
+            }
+        }
+    }
+
+    #[tokio::test]
+    async fn fails_when_invalid_range() {
+        let (stream1, stream2) = duplex(4096);
+        let mut invalid_range = supported_protocol_range();
+        invalid_range.0 = invalid_range.1 + 1;
+        let negotiation1 = negotiate_protocol_version(stream1, invalid_range.clone()).fuse();
+        pin_mut!(negotiation1);
+        let negotiation2 = negotiate_protocol_version(stream2, invalid_range.clone()).fuse();
+        pin_mut!(negotiation2);
+        for _ in 0..2 {
+            tokio::select! {
+                result = &mut negotiation1 => incorrect_negotiation(result, ProtocolNegotiationError::InvalidRange(invalid_range.clone())),
+                result = &mut negotiation2 => incorrect_negotiation(result, ProtocolNegotiationError::InvalidRange(invalid_range.clone())),
+            }
+        }
+    }
+
+    #[tokio::test]
+    async fn fails_when_connection_droped() {
```

Comment:
droped -> dropped

Reply:
Fixed, plus added a small trait requirement which we will actually need later.